cfgov/jobmanager/migrations/0012_jobs_have_one_region.py

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import json
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+HQ_REGION_ABBREVIATION = 'HQ'
+
+
+def check_prerequisites(apps, schema_editor):
+    JobListingPage = apps.get_model('jobmanager', 'JobListingPage')
+    Location = apps.get_model('jobmanager', 'Location')
+
+    # Assert all jobs have at least one region.
+    for page in JobListingPage.objects.all():
+        if not page.regions.exists():
+            raise RuntimeError('cannot migrate page without region')
+
+    # Assert that the headquarters region exists.
+    Location.objects.get(region=HQ_REGION_ABBREVIATION)
+
+
+def create_jobregions(apps, schema_editor):
+    JobRegion = apps.get_model('jobmanager', 'JobRegion')
+    Location = apps.get_model('jobmanager', 'Location')
+
+    for location in Location.objects.all():
+        JobRegion.objects.create(
+            abbreviation=location.region,
+            name=location.region_long
+        )
+
+
+def create_locations(apps, schema_editor):
+    JobRegion = apps.get_model('jobmanager', 'JobRegion')
+    Location = apps.get_model('jobmanager', 'Location')
+
+    for job_region in JobRegion.objects.all():
+        Location.objects.create(
+            description=job_region.name,
+            region=job_region.abbreviation,
+            region_long=job_region.name
+        )
+
+
+def multiple_regions_to_single_region(apps, schema_editor):
+    JobListingPage = apps.get_model('jobmanager', 'JobListingPage')
+    JobRegion = apps.get_model('jobmanager', 'JobRegion')
+
+    hq_region = JobRegion.objects.get(abbreviation=HQ_REGION_ABBREVIATION)
+    for job_listing in JobListingPage.objects.all():
+        if 1 < job_listing.regions.count():
+            new_region = hq_region
+        else:
+            old_region = job_listing.regions.get().region
+            new_region = JobRegion.objects.get(abbreviation=old_region.region)
+
+        new_region.job_listings.add(job_listing)
+
+        for revision in job_listing.revisions.all():
+            content = json.loads(revision.content_json)
+            content.pop('regions', None)
+            content['region'] = new_region.pk
+            revision.content_json = json.dumps(content)
+            revision.save()
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('jobmanager', '0011_delete_fellowshipupdatelist'),
+    ]
+
+    operations = [
```

Comment:
This migration is a bit complicated, so here's what it's doing:

1. Runs `check_prerequisites` to verify two database assumptions: all jobs have at least one region and a region named "HQ" exists.
1. Creates the new region model as `jobmanager.JobRegion`.
1. Copies over existing `jobmanager.Location` model instances into new `jobmanager.JobRegion`s.
1. Adds a new `region` field to the `JobListingPage` page type, making it potentially null.
1. Runs `multiple_regions_to_single_region` to update all existing `JobListingPage` pages, to copy their (potentially multiple) region(s) from `Location` to `JobRegion` instances. This includes not only updating the page model but also all historical revisions. ಠ_ಠ
1. Updates the `JobListingPage` page type to make its `region` field not nullable.
1. Deletes the now-unneeded `RegionPanel` and `Location` models.